modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/custom/form/FormView.es.js

```diff
@@ -99,11 +98,10 @@ const useFormSubmit = ({apiRef, containerRef}) => {
 
 	useEffect(() => {
 		let onHandle;
-
-		let submitHandle;
+		let form;
```

Comment:
`const` is usually preferable to `let`, if you can bend the code to allow it, because it allows the reader to go from top to bottom without having to keep track of possible mutations along the way.

In this case, the following might make this more readable:

```js
useEffect(() => {
    if (containerRef.current) {
        const form = getFormNode(containerRef.current);

        if (form) {
            const onHandle = Liferay.on(
                'submitForm',
                (event) => {
                    if (event.form && event.form.getDOM() === form) {
                        event.preventDefault();
                    }
                },
                this
            );

            form.addEventListener('submit', handleFormSubmitted);

            return () => {
                onHandle.detach();
                form.removeEventListener('submit', handleFormSubmitted);
            };
        }
    }
}, [containerRef, handleFormSubmitted]);
```

Does that seem better to you? Note also how we avoid the cost of returning a function in the event that we don't need to, and we are able to drop some `if` conditionals.

Another thing I would do is actually roll this in together with your other PR, so you don't have to wait for two full cycles of: wait for CI to pass, forward to Brian, wait for Brian to merge, potentially wait for upstream to sync from Brian's fork, rebase, wait for CI to pass etc...

Reply:
Disclaimer: I am not 100% confident of that code so don't trust it without checking it. It's possible I might me leaking event listeners. I don't _think_ I am, but I might be. 😂

Reply:
1) your recommendation looks much better, it occured to me, but I blindly trusted whoever wrote the initial code that just moving the `return` statement into the `if` check wouldn't work. Still rusty with `useEffect` so didn't wanna spend more time than needed to get eyes on this.

2) I'll combine the 2 PRs into one, that's a good idea, wasn't sure if that was proper JIRA/task etiquette

3) Got any tips on debugging event handlers/listeners? Google said to add a boolean check, some other dude recommended a 3rd party software. What do you do when you can't log the listener? Primarily interested in testing if it's been properly removed.

Reply:
3. I don't think you need to actually test it. Just try running it and see it work. And analyze it to convince yourself that it's not buggy. The thing about leaking event listeners is that, if it happens, it could be intermittent (dependent on the fragile set of circumstances that determine what the `if` conditions evaluate to), so the _analysis_ is probably more important than actually _testing_.

> Still rusty with `useEffect` so didn't wanna spend more time than needed to get eyes on this.

Being confident about how `useEffect` works would actually be a very valuable investment for you because it is one of the trickiest parts of React. If you don't really get it, it will bite you again and again.

In this case, like I said, I don't think I'm leaking a listener, and here's why:

- `useEffect` will run on ever render.
- The callback passed into `useEffect` will run whenever anything in the dependency array changes.
- In the callback, unless both `containerRef.current` and `form` are both present, we do nothing, which means there is nothing to clean up.
- If we do set something up (ie. because we have both `containerRef.current` and `form`), then we are guaranteed to do our clean-up on the next unmount or update (whichever comes first), which means there is no way we are going to leak a listener.

The middle ground you've gone with in [your other pull](https://github.com/liferay-frontend/liferay-portal/pull/935) still has the `let` and the repeated `if` checks, leading to the readability issues. 

```js
	useEffect(() => {
		let onHandle;

		if (containerRef.current) {
			const form = getFormNode(containerRef.current);

			if (form) {
				onHandle = Liferay.on(
					'submitForm',
					(event) => {
						if (event.form && event.form.getDOM() === form) {
							event.preventDefault();
						}
					},
					this
				);

				form.addEventListener('submit', handleFormSubmitted);
			}

			return () => {
				if (onHandle) {
					onHandle.detach();
				}

				if (form) {
					form.removeEventListener('submit', handleFormSubmitted);
				}
			};
		}
	}, [containerRef, handleFormSubmitted]);
```

The thing I don't like about that is in your `return` you have those two `if` checks. But note that they will always both pass or both fail; they can never pass alone. So you're checking two separate things that both equate to effectively the same thing. It will work, but it just adds to the cognitive load of reading it.